servers/src/main/java/tachyon/worker/block/TieredBlockStore.java

```diff
@@ -0,0 +1,432 @@
+    }
+    mMetaManager.abortTempBlockMeta(tempBlockMeta);
+  }
+
+  private void moveBlockNoLock(long blockId, BlockStoreLocation newLocation) throws IOException {
+    if (mMetaManager.hasTempBlockMeta(blockId)) {
+      throw new IOException("Failed to move block " + blockId + ": block is uncommited");
+    }
+    BlockMeta blockMeta = mMetaManager.getBlockMeta(blockId);
+    String srcPath = blockMeta.getPath();
+    blockMeta = mMetaManager.moveBlockMeta(blockMeta, newLocation);
+    String destPath = blockMeta.getPath();
+
+    if (!new File(srcPath).renameTo(new File(destPath))) {
+      throw new IOException("Failed to move block " + blockId + ": cannot rename from " + srcPath
+          + " to " + destPath);
+    }
+  }
+
+  // TODO: refactor this method: currently, it is shared by code path of both remove, eviction
+  // and remove temp data. Let us separate it.
+  private void removeBlockNoLock(long userId, long blockId) throws IOException {
+    // Delete metadata of the block---no matter it is a temp block.
+    String filePath;
+    if (mMetaManager.hasTempBlockMeta(blockId)) {
+      TempBlockMeta tempBlockMeta = mMetaManager.getTempBlockMeta(blockId);
+      mMetaManager.abortTempBlockMeta(tempBlockMeta);
+      filePath = tempBlockMeta.getPath();
+    } else if (mMetaManager.hasBlockMeta(blockId)) {
+      BlockMeta blockMeta = mMetaManager.getBlockMeta(blockId);
+      mMetaManager.removeBlockMeta(blockMeta);
+      filePath = blockMeta.getPath();
+    } else {
+      throw new IOException("Failed to move block " + blockId + ": block is not found");
+    }
+
+    // Delete the data of the block on "disk"
+    if (!new File(filePath).delete()) {
+      throw new IOException("Failed to remove block " + blockId + ": cannot delete " + filePath);
+    }
+  }
+
+  // This method must be guarded by WRITE lock of mEvictionLock
+  private void freeSpaceInternal(long userId, long availableBytes, BlockStoreLocation location)
+      throws IOException {
+    EvictionPlan plan = mEvictor.freeSpace(availableBytes, location);
+    // Absent plan means failed to evict enough space.
+    if (plan == null) {
+      throw new IOException("Failed to free space: no eviction plan by evictor");
+    }
+
+    // Step1: remove blocks to make room.
+    for (long blockId : plan.toEvict()) {
+      long lockId = mLockManager.lockBlock(userId, blockId, BlockLockType.WRITE);
+      try {
+        removeBlockNoLock(userId, blockId);
+        for (BlockStoreEventListener listener : mBlockStoreEventListeners) {
+          listener.onRemoveBlockByWorker(userId, blockId);
+        }
+      } catch (IOException e) {
+        throw new IOException("Failed to free space: cannot evict block " + blockId);
+      } finally {
+        mLockManager.unlockBlock(lockId);
+      }
+    }
+    // Step2: transfer blocks among tiers.
+    for (Pair<Long, BlockStoreLocation> entry : plan.toMove()) {
```

Comment:
Should ensure we move to the lowest tier first in order to guarantee we have space.